substrate/primitives/trie/src/hasher_random_state.rs

```diff
@@ -0,0 +1,123 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utility module to use a custom random state for HashMap and friends
+//! in a no_std environment.
+
+use core::{
+	hash::Hasher as CoreHasher,
+	sync::atomic::{AtomicUsize, Ordering},
+};
+
+use core::hash::BuildHasher;
+use hashbrown::DefaultHashBuilder;
+
+// Extra randomness to be used besides the one provided by the `DefaultHashBuilder`.
+static EXTRA_RANDOMNESS: AtomicUsize = AtomicUsize::new(0x082efa98);
+
+/// Adds extra randomness to be used by all new instances of RandomState.
+pub fn add_extra_randomness(extra_randomness: usize) {
+	EXTRA_RANDOMNESS.store(extra_randomness, Ordering::Relaxed);
+}
+
+/// A wrapper around `DefaultHashBuilder` that adds extra randomness to the hashers it creates.
+#[derive(Copy, Clone, Debug)]
+pub struct RandomState {
+	default: DefaultHashBuilder,
+	extra_randomness: usize,
+}
+
+impl Default for RandomState {
+	#[inline(always)]
+	fn default() -> Self {
+		RandomState {
+			// DefaultHashBuild already uses a random seed, so we use that as the base.
+			default: DefaultHashBuilder::default(),
+			extra_randomness: EXTRA_RANDOMNESS.load(Ordering::Relaxed),
+		}
+	}
+}
+
+impl BuildHasher for RandomState {
+	type Hasher = <DefaultHashBuilder as BuildHasher>::Hasher;
+
+	#[inline(always)]
+	fn build_hasher(&self) -> Self::Hasher {
+		let mut hasher = self.default.build_hasher();
+		hasher.write_usize(self.extra_randomness);
```

Comment:
Wouldn't it make more sense to use a bigger seed? Wouldn't this improve the quality of the randomness? (No expert)

Reply:
How big is the seed here?   128 bits?  A priori, we'd probably gain nothing beyond that given the tiny output size, except we use hashbrown here, which winds up weaker than sip hash.  We cannot rule out something that exploits the structure of hash brown.